drivers/gs.go

```diff
@@ -0,0 +1,162 @@
+package drivers
+
+import (
+	"crypto"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"time"
+
+	"github.com/livepeer/go-livepeer/net"
+)
+
+type (
+	gsKeyJSON struct {
+		Type                    string `json:"type,omitempty"`
+		ProjectID               string `json:"project_id,omitempty"`
+		PrivateKeyID            string `json:"private_key_id,omitempty"`
+		PrivateKey              string `json:"private_key,omitempty"`
+		ClientEmail             string `json:"client_email,omitempty"`
+		ClientID                string `json:"client_id,omitempty"`
+		AuthURI                 string `json:"auth_uri,omitempty"`
+		TokenURI                string `json:"token_uri,omitempty"`
+		AuthProviderX509CertURL string `json:"auth_provider_x509_cert_url,omitempty"`
+		ClientX509CertURL       string `json:"client_x509_cert_url,omitempty"`
+	}
+
+	gsSigner struct {
+		jsKey     *gsKeyJSON
+		parsedKey *rsa.PrivateKey
+	}
+
+	gsOS struct {
+		s3OS
+		gsSigner *gsSigner
+	}
+
+	gsSession struct {
+		s3Session
+	}
+)
+
+func gsHost(bucket string) string {
+	return fmt.Sprintf("https://%s.storage.googleapis.com", bucket)
+}
+
+func gsParseKey(key []byte) (*rsa.PrivateKey, error) {
+	if block, _ := pem.Decode(key); block != nil {
+		key = block.Bytes
+	}
+	parsedKey, err := x509.ParsePKCS8PrivateKey(key)
+	if err != nil {
+		parsedKey, err = x509.ParsePKCS1PrivateKey(key)
+		if err != nil {
+			return nil, err
+		}
+	}
+	parsed, ok := parsedKey.(*rsa.PrivateKey)
+	if !ok {
+		return nil, errors.New("oauth2: private key is invalid")
+	}
+	return parsed, nil
+}
+
+func NewGoogleDriver(bucket, keyFileName string) (OSDriver, error) {
+	os := &gsOS{
+		s3OS: s3OS{
+			host:   gsHost(bucket),
+			bucket: bucket,
+		},
+	}
+	rawFile, err := ioutil.ReadFile(keyFileName)
+	if err != nil {
+		return nil, err
+	}
+	var gsKey gsKeyJSON
+	if err := json.Unmarshal(rawFile, &gsKey); err != nil {
+		return nil, err
+	}
+	parsedKey, err := gsParseKey([]byte(gsKey.PrivateKey))
+	if err != nil {
+		return nil, err
+	}
+	os.gsSigner = &gsSigner{
+		jsKey:     &gsKey,
+		parsedKey: parsedKey,
+	}
+	return os, nil
+}
+
+func (os *gsOS) NewSession(path string) OSSession {
+	var policy, signature = gsCreatePolicy(os.gsSigner, os.bucket, os.region, path)
+
+	sess := os.s3OS.NewSession(path).(*s3Session)
+	sess.credential = os.gsSigner.clientEmail()
+	sess.host = gsHost(os.bucket)
+	sess.policy = policy
+	sess.signature = signature
+	return &gsSession{s3Session: *sess}
+}
+
+func (os *gsSession) SaveData(name string, data []byte) (string, error) {
+	fields := map[string]string{
+		"GoogleAccessId": os.credential,
+		"signature":      os.signature,
+	}
+	return os.saveData(name, data, fields)
+}
+
+func (os *gsSession) GetInfo() *net.OSInfo {
+	oi := os.s3Session.GetInfo()
+	oi.StorageType = net.OSInfo_GOOGLE
```

Comment:
This works out nicely (along with the use of `newS3Session`) because these functions don't require changes within the s3 driver to accommodate google storage. Looks great.